### PR TITLE
Reconcile grad-accum loss normalization in _unsloth_get_batch_samples

### PR DESCRIPTION
**Pull request by danielhanchen**
Three components each decide "is this loss already token-count normalised", and they use different predicates:

| component | predicate |
|---|---|
| `_unsloth_get_batch_samples` | forward signature takes `**kwargs` |
| our `unsloth_fused_ce_loss`, TRL's `_chunked_cross_entropy_loss` | `num_items_in_batch is not None` |
| `Trainer.training_step` | `self.model_accepts_loss_kwargs` |

When a model class sets `accepts_loss_kwargs = False` (gemma3, gemma4, qwen-vl, paligemma, glm4v) and the loss still divides by the token count, two of the three normalise. Loss and gradients are scaled by 1/GA, so the effective learning rate is GA times too small. Nothing raises.

This is masked on the default path, because `apply_accepts_loss_kwargs_fix` forces the flag True whenever the forward is compiled. It is reachable with `UNSLOTH_COMPILE_DISABLE=1`, which is a common first step when debugging something else.

Measured on real `unsloth/gemma-3-4b-it-unsloth-bnb-4bit`, no forcing, only the env var set:

| | GA=1 | GA=2 | GA=4 |
|---|---|---|---|
| before, grad-norm | 1.041674 | 0.520837 | 0.260419 |
| after, grad-norm | 1.041674 | 1.041674 | 1.041674 |

## Change

Reconcile the two predicates before returning from `_unsloth_get_batch_samples`. Two cases need opposite treatment:

- The `False` came from the model class. The loss really is token normalised, so set `model_accepts_loss_kwargs = True` and let `training_step` skip its own divide. This keeps the token-weighted normalisation that the upstream grad-accum fix exists to provide.
- A trainer set `False` on purpose to enable the divide. Return no count instead, which is exactly what stock transformers would have passed, since `Trainer._get_num_items_in_batch` only counts when `self.model_accepts_loss_kwargs or self.compute_loss_func is not None`.

The second case is load-bearing. Thirteen TRL trainers assign `self.model_accepts_loss_kwargs = False` in `__init__` with the comment "to enable scaling": DPO, KTO, GRPO, RLOO, Reward, CPO, ORPO, BCO, Distillation, SDPO, SSD, SDFT and AsyncGRPO. Blindly forcing the flag True would silently break every one of them, so the two are told apart by looking for that assignment in the trainer's own class hierarchy, cached per class. If the source cannot be read the code falls back to the conservative branch, which can only ever reproduce stock transformers behaviour.

## Verification

GA-invariance on gemma-3-4b, `UNSLOTH_COMPILE_DISABLE=1`, grad-norm ratios versus GA=1:

| scenario | GA=1 | GA=2 | GA=4 |
|---|---|---|---|
| this change alone, TRL default | x1.000 | x1.000 | x1.000 |
| this change plus explicit `loss_type="chunked_nll"` | x1.000 | - | x1.000 |

New `tests/test_loss_normalization_reconcile.py`, 20 tests, CPU only, no downloads: the dispatch table for every flag and count combination, one parametrised case per opt-out TRL trainer, SFT and base `Trainer` correctly not flagged, per-class caching, regex spacing variants, and a drift check that fails if transformers stops keying off the flag.

Downstream regression: the full `unsloth` `tests/version_compat/` suite passes at 1697 passed, 182 skipped with this change installed.

## Relationship to unslothai/unsloth#8261

That PR pins the generated SFT default back to `loss_type="nll"`, which sidesteps the same bug for default SFT runs and restores our fused CE. This one fixes the underlying predicate disagreement, so it also covers an explicit `loss_type="chunked_nll"`, which the other PR leaves exposed (reproducible back to TRL 1.4.0). They are independent and either fixes default SFT on its own; I would take both, since they address different layers.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Marking this draft. Re-measured after unslothai/unsloth#8261 landed, and this change is wrong in its current form.

Real `unsloth/gemma-3-4b-it-unsloth-bnb-4bit`, grad-norm at GA=1 vs GA=4, only this commit varied:

**With `UNSLOTH_COMPILE_DISABLE=1`:**

| loss_type | without this PR | with this PR |
|---|---|---|
| `nll` (the default after #8261) | 1.041674 / 1.041674, correct | 1.041674 / **4.166698, 4x too large** |
| explicit `chunked_nll` | 1.041674 / 0.260419, the 1/GA bug | 1.041674 / 1.041674, correct |

**With compile enabled (production default):** identical with and without this PR, GA-invariant either way, since `apply_accepts_loss_kwargs_fix` already forces the flag True so the reconcile never fires.

The dispatch assumes that a returned `num_items_in_batch` means the active loss divides by it, and flips `model_accepts_loss_kwargs` to True on that basis. Under `nll` the loss does not consume the count, so suppressing the grad-accum divide in `training_step` leaves gradients GA times too large. That trades one silent scaling bug for its mirror image, which is exactly the failure the docstring in this PR warns about.

The "this change alone, TRL default = x1.000" row in the description above is stale: it was measured when TRL's default was still `chunked_nll`, before #8261 pinned it back to `nll`.

Still worth fixing, since #8261 does not subsume this. An explicit `loss_type="chunked_nll"` is still 1/GA-scaled on merged main. The fix is to gate the flip on whether the active loss actually consumes the count, using the resolved `loss_type` and whether the fused CE is installed, rather than on where the `False` came from, plus a regression case for default `nll` with compile disabled.

**Comment by danielhanchen** [comment]
Fixed and re-measured. Taking this out of draft.

The dispatch now keys off whether the **active loss actually consumes** `num_items_in_batch`, rather than off where the `False` came from. It reuses the condition already encoded in `_chunked_loss_bypasses_forward`: TRL's chunked-CE lm_head patch present, liger not forcing `nll`, and a resolved `loss_type` of `chunked_nll` or a still-unset `None`.

Grad-norm on `unsloth/gemma-3-4b-it-unsloth-bnb-4bit`, GA=1 vs GA=4:

| scenario | before this commit | after |
|---|---|---|
| `nll`, compile disabled | 1.041674 / **4.166698** | 1.041674 / 1.041674 |
| explicit `chunked_nll`, compile disabled | 1.041674 / 1.041674 | 1.041674 / 1.041674 |
| `nll`, compile enabled (production default) | 1.052273 / 1.052273 | 1.052273 / 1.052273 |

Three regression cases added: plain `nll`, `use_liger_kernel`, and an unresolved `loss_type=None`. The earlier tests all used synthetic trainers with no `args`, so `loss_type` was absent from the test space entirely, which is why this got through. 23 tests pass.

The description at the top of this PR still has one stale row, the "this change alone, TRL default = x1.000" line, measured when TRL's default was still `chunked_nll`. The table above supersedes it.

**Comment by danielhanchen** [comment]
Rewritten much smaller, and it now resolves the review findings by construction rather than by patching around them. Net 9 insertions against 70 deletions.

The whole change is now the stock predicate:

```python
if (num_items_in_batch is not None
        and not getattr(self, "model_accepts_loss_kwargs", True)
        and getattr(self, "compute_loss_func", None) is None):
    num_items_in_batch = None
```

That is exactly what `Trainer._get_num_items_in_batch` gates on upstream. It is safe because every loss that divides by the count falls back to a mean when it is None, TRL's `_chunked_cross_entropy_loss` and our own `unsloth_fused_ce_loss` included, so `training_step`'s grad-accum division becomes the single correct normalisation.

Gone with it: the source introspection, the per-class cache, the thirteen-entry trainer table, and any assignment to `model_accepts_loss_kwargs`.

## Review findings

- **Flip the flag only when the active loss consumes the count.** Correct, and this was the more serious of the two. I measured 4.166698 against 1.041674 at GA=4 under `nll` once #8261 pinned the default back. The rewrite never flips the flag at all.
- **Preserve the sequence-parallel loss-scaling opt-out.** Correct. Nothing assigns the flag now, so the `Trainer.__init__` DeepSpeed opt-out and the thirteen TRL trainers that disable it deliberately are all untouched. Suppressing the count is the behaviour they already expect.
- **Drop RewardTrainer from the deliberate-disable table.** Correct, and verified: on TRL 0.24.0 `RewardTrainer.__init__` does not assign the flag, on 1.9.2 it does. The table is gone.
- **Fail instead of skipping when loss_utils cannot import.** Fair in principle. The suite is now source-level assertions with a real negative control rather than synthetic trainers.

## Verification

Grad-norm on `unsloth/gemma-3-4b-it-unsloth-bnb-4bit`, GA=1 vs GA=4:

| scenario | stock | with this PR |
|---|---|---|
| explicit `chunked_nll`, compile disabled | 1.041674 / **0.260419** | 1.041674 / 1.041674 |
| `nll` (the default after #8261), compile disabled | 1.041674 / 1.041674 | 1.041674 / 1.041674 |
| `nll`, compile enabled (production default) | 1.052273 / 1.052273 | 1.052273 / 1.052273 |

The default path is already correct on merged main, so this only closes the remaining hole: an explicit `loss_type="chunked_nll"` is still 1/GA-scaled without it.

Six tests in `tests/test_loss_normalization_contract.py`, run from outside the repo so the installed package is not shadowed: 1 fails on stock, all 6 pass with the fix.

Both earlier tables in this thread are superseded by the one above.

File: tests/test_loss_normalization_reconcile.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+import importlib
+import re
+
+import pytest
+
+
+def _fn(name):
+    mod = pytest.importorskip("unsloth_zoo.loss_utils")
+    fn = getattr(mod, name, None)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tests/test_loss_normalization_reconcile.py]: Not taking this one. The file it was raised against is gone as of 287ad6a2, and the pattern it objects to is the convention across this suite: `tests/conftest.py` pre-loads `unsloth_zoo.device_type` so the package imports without an accelerator, and `importorskip("unsloth_zoo...")` is used the same way in the dataset_utils, dataset_num_proc and compiler tests. In CI the Repo tests job installs unsloth (`--no-deps`) to satisfy the `find_spec("unsloth")` guard at `unsloth_zoo/__init__.py:128`, so these tests execute rather than skip: on the current head the file reports 7 passed, 1 skipped, the single skip being the TRL chunked CE probe on a TRL without `_chunked_cross_entropy_loss`.

Making this one file import bare would only make it the first to fail for an environment reason unrelated to the guard.

File: unsloth_zoo/loss_utils.py [hunk 2]
```diff
@@ -335,6 +336,18 @@ def _unsloth_get_batch_samples(self, epoch_iterator, num_batches, device = None,
...
+    # We decide num_items_in_batch from the forward signature, but training_step
+    # divides by grad-accum based on self.model_accepts_loss_kwargs. Returning a
+    # count while that flag is False normalises the loss twice (TRL's chunked_nll
+    # and our fused CE both divide by it), silently scaling loss and grads by 1/GA.
+    # Mirror stock Trainer._get_num_items_in_batch: only count when someone will
+    # consume it. Every such loss falls back to a mean when it is None.
+    if (num_items_in_batch is not None
+            and not getattr(self, "model_accepts_loss_kwargs", True)
+            and getattr(self, "compute_loss_func", None) is None):
+        num_items_in_batch = None
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/loss_utils.py:349]: Not taking this one, because the alternative is the bug this PR removes.

The observation is right on its own terms: with the count cleared, ragged microbatches give a mean of means rather than a global token mean. But that is stock transformers behaviour for this state, not something the guard introduces. `Trainer._get_num_items_in_batch` counts only when `self.model_accepts_loss_kwargs or self.compute_loss_func is not None`, and `training_step` divides by `current_gradient_accumulation_steps` when `(not self.model_accepts_loss_kwargs or num_items_in_batch is None) and self.compute_loss_func is None`. Both lines are byte identical in 4.57.6 and 5.5.0. The two are complementary, so the state this item asks us to keep, a count alongside a False flag and no compute_loss_func, is one stock never produces.

Keeping the count there means dividing twice, once by the global count in the loss and again by GA in `training_step`. Measured on a CPU harness with a token normalising loss, GA=4, four equal-length microbatches, gradient of `lm_head.weight` against the same data as one batch:

| | ratio versus single batch |
|---|---|
| count kept, flag False | x0.25 |
| count cleared (this PR) | x1.000, max abs diff 3.7e-8 |

The only way to keep the count and not divide twice is to set `model_accepts_loss_kwargs = True`, which is what e4817c01 and e79eeb16 did and what the earlier items on this PR correctly rejected: Gemma3's native loss is a plain `nn.CrossEntropyLoss()` mean that ignores the count, so flipping the flag there scales gradients up by GA instead.

Both tests are in b30e40e4, and I confirmed each fails with the guard reverted.
